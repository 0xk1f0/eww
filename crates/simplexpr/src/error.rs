use crate::{
    dynval,
    parser::lexer::{self, LexicalError},
};
use eww_shared_util::Span;

pub type Result<T> = std::result::Result<T, Error>;
#[derive(thiserror::Error, Debug)]
pub enum Error {
    #[error("Parse error: {source}")]
    ParseError { file_id: usize, source: lalrpop_util::ParseError<usize, lexer::Token, lexer::LexicalError> },

    #[error("Type error: {0}")]
    ConversionError(#[from] dynval::ConversionError),

    #[error("{1}")]
    Spanned(Span, Box<Error>),

    #[error(transparent)]
    Eval(#[from] crate::eval::EvalError),

    #[error(transparent)]
    Other(#[from] Box<dyn std::error::Error + Send + Sync>),
}

impl Error {
    pub fn from_parse_error(file_id: usize, err: lalrpop_util::ParseError<usize, lexer::Token, lexer::LexicalError>) -> Self {
        Error::ParseError { file_id, source: err }
    }

    pub fn at(self, span: Span) -> Self {
        Self::Spanned(span, Box::new(self))
    }

    pub fn get_span(&self) -> Option<Span> {
        match self {
            Self::ParseError { file_id, source } => get_parse_error_span(*file_id, source),
            Self::Spanned(span, _) => Some(*span),
            Self::Eval(err) => err.span(),
            Self::ConversionError(err) => err.span(),
            _ => None,
        }
    }
}

fn get_parse_error_span(
    file_id: usize,
    err: &lalrpop_util::ParseError<usize, lexer::Token, lexer::LexicalError>,
) -> Option<Span> {
    match err {
        lalrpop_util::ParseError::InvalidToken { location } => Some(Span(*location, *location, file_id)),
        lalrpop_util::ParseError::UnrecognizedEOF { location, expected: _ } => Some(Span(*location, *location, file_id)),
        lalrpop_util::ParseError::UnrecognizedToken { token, expected: _ } => Some(Span(token.0, token.2, file_id)),
        lalrpop_util::ParseError::ExtraToken { token } => Some(Span(token.0, token.2, file_id)),
        lalrpop_util::ParseError::User { error: LexicalError(l, r) } => Some(Span(*l, *r, file_id)),
    }
}

#[macro_export]
macro_rules! spanned {
    ($err:ty, $span:expr, $block:expr) => {{
        let span = $span;
        let result: Result<_, $err> = try { $block };
        result.at(span)
    }};
}
