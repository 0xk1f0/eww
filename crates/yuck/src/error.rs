use crate::{
    config::{attributes::AttrError, config::Include, validate::ValidationError},
    parser::{
        ast::{Ast, AstType},
        lexer, parse_error,
    },
};
use codespan_reporting::{diagnostic, files};
use eww_shared_util::{AttrName, Span, VarName};
use simplexpr::dynval;
use thiserror::Error;

pub type AstResult<T> = Result<T, AstError>;

#[derive(Debug, Error)]
pub enum AstError {
    #[error("Unknown toplevel declaration `{1}`")]
    UnknownToplevel(Span, String),
    #[error("Expected another element, but got nothing")]
    MissingNode(Span),
    #[error("Too many elements, must be exactly {1}")]
    TooManyNodes(Span, i32),

    #[error("Wrong type of expression: Expected {1} but got {2}")]
    WrongExprType(Span, AstType, AstType),
    #[error("Expected to get a value, but got {1}")]
    NotAValue(Span, AstType),
    #[error("Expected element {1}, but read {2}")]
    MismatchedElementName(Span, String, String),

    #[error("Included file not found {}", .0.path)]
    IncludedFileNotFound(Include),

    #[error(transparent)]
    ConversionError(#[from] dynval::ConversionError),

    #[error("{1}")]
    Other(Option<Span>, Box<dyn std::error::Error + Sync + Send + 'static>),

    #[error(transparent)]
    AttrError(#[from] AttrError),

    #[error(transparent)]
    ValidationError(#[from] ValidationError),

    #[error("Parse error: {source}")]
    ParseError { file_id: Option<usize>, source: lalrpop_util::ParseError<usize, lexer::Token, parse_error::ParseError> },
}

// static_assertions::assert_impl_all!(AstError: Send, Sync);
// static_assertions::assert_impl_all!(dynval::ConversionError: Send, Sync);
// static_assertions::assert_impl_all!(lalrpop_util::ParseError < usize, lexer::Token, parse_error::ParseError>: Send, Sync);

impl AstError {
    pub fn get_span(&self) -> Option<Span> {
        match self {
            AstError::UnknownToplevel(span, _) => Some(*span),
            AstError::MissingNode(span) => Some(*span),
            AstError::WrongExprType(span, ..) => Some(*span),
            AstError::NotAValue(span, ..) => Some(*span),
            AstError::MismatchedElementName(span, ..) => Some(*span),
            AstError::AttrError(err) => Some(err.span()),
            AstError::Other(span, ..) => *span,
            AstError::ConversionError(err) => err.value.span().map(|x| x.into()),
            AstError::IncludedFileNotFound(include) => Some(include.path_span),
            AstError::TooManyNodes(span, ..) => Some(*span),
            AstError::ValidationError(error) => None, // TODO none here is stupid
            AstError::ParseError { file_id, source } => file_id.and_then(|id| get_parse_error_span(id, source)),
        }
    }

    pub fn from_parse_error(
        file_id: usize,
        err: lalrpop_util::ParseError<usize, lexer::Token, parse_error::ParseError>,
    ) -> AstError {
        AstError::ParseError { file_id: Some(file_id), source: err }
    }
}

fn get_parse_error_span(
    file_id: usize,
    err: &lalrpop_util::ParseError<usize, lexer::Token, parse_error::ParseError>,
) -> Option<Span> {
    match err {
        lalrpop_util::ParseError::InvalidToken { location } => Some(Span(*location, *location, file_id)),
        lalrpop_util::ParseError::UnrecognizedEOF { location, expected } => Some(Span(*location, *location, file_id)),
        lalrpop_util::ParseError::UnrecognizedToken { token, expected } => Some(Span(token.0, token.2, file_id)),
        lalrpop_util::ParseError::ExtraToken { token } => Some(Span(token.0, token.2, file_id)),
        lalrpop_util::ParseError::User { error } => match error {
            parse_error::ParseError::SimplExpr(span, error) => *span,
            parse_error::ParseError::LexicalError(span) => Some(*span),
        },
    }
}

pub trait OptionAstErrorExt<T> {
    fn or_missing(self, span: Span) -> Result<T, AstError>;
}
impl<T> OptionAstErrorExt<T> for Option<T> {
    fn or_missing(self, span: Span) -> Result<T, AstError> {
        self.ok_or(AstError::MissingNode(span))
    }
}
