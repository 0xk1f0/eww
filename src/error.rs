use crate::parser::{
    ast::{Ast, AstType, Span},
    lexer, parse_error,
};
use codespan_reporting::{diagnostic, files};
use thiserror::Error;

pub type AstResult<T> = Result<T, AstError>;

#[derive(Debug, Error)]
pub enum AstError {
    #[error("Definition invalid")]
    InvalidDefinition(Option<Span>),
    #[error("Expected a {1}, but got nothing")]
    MissingNode(Option<Span>, AstType),
    #[error("Wrong type of expression: Expected {1} but got {2}")]
    WrongExprType(Option<Span>, AstType, AstType),

    #[error("Parse error: {source}")]
    ParseError { file_id: Option<usize>, source: lalrpop_util::ParseError<usize, lexer::Token, parse_error::ParseError> },
}

impl AstError {
    pub fn get_span(&self) -> Option<Span> {
        match self {
            AstError::InvalidDefinition(span) => *span,
            AstError::MissingNode(span, _) => *span,
            AstError::WrongExprType(span, ..) => *span,
            AstError::ParseError { file_id, source } => file_id.and_then(|id| get_parse_error_span(id, source)),
        }
    }

    pub fn from_parse_error(
        file_id: usize,
        err: lalrpop_util::ParseError<usize, lexer::Token, parse_error::ParseError>,
    ) -> AstError {
        AstError::ParseError { file_id: Some(file_id), source: err }
    }
}

fn get_parse_error_span(
    file_id: usize,
    err: &lalrpop_util::ParseError<usize, lexer::Token, parse_error::ParseError>,
) -> Option<Span> {
    match err {
        lalrpop_util::ParseError::InvalidToken { location } => Some(Span(*location, *location, file_id)),
        lalrpop_util::ParseError::UnrecognizedEOF { location, expected } => Some(Span(*location, *location, file_id)),
        lalrpop_util::ParseError::UnrecognizedToken { token, expected } => Some(Span(token.0, token.2, file_id)),
        lalrpop_util::ParseError::ExtraToken { token } => Some(Span(token.0, token.2, file_id)),
        lalrpop_util::ParseError::User { error } => match error {
            parse_error::ParseError::SimplExpr(span, error) => *span,
            parse_error::ParseError::LexicalError(span) => Some(*span),
        },
    }
}

pub fn spanned(span: Span, err: impl Into<AstError>) -> AstError {
    use AstError::*;
    match err.into() {
        AstError::InvalidDefinition(None) => AstError::InvalidDefinition(Some(span)),
        AstError::MissingNode(None, x) => AstError::MissingNode(Some(span), x),
        AstError::WrongExprType(None, x, y) => AstError::WrongExprType(Some(span), x, y),
        x => x,
    }
}

pub trait OptionAstErrorExt<T> {
    fn or_missing(self, t: AstType) -> Result<T, AstError>;
}
impl<T> OptionAstErrorExt<T> for Option<T> {
    fn or_missing(self, t: AstType) -> Result<T, AstError> {
        self.ok_or(AstError::MissingNode(None, t))
    }
}

pub trait AstResultExt<T> {
    fn at(self, span: Span) -> Result<T, AstError>;
}

impl<T, E: Into<AstError>> AstResultExt<T> for Result<T, E> {
    fn at(self, span: Span) -> Result<T, AstError> {
        self.map_err(|err| spanned(span, err))
    }
}

#[macro_export]
macro_rules! spanned {
    ($span:expr, $block:expr) => {{
        let span = $span;
        let result: Result<_, AstError> = try { $block };
        result.at(span)
    }};
}
