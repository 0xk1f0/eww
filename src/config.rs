use std::{collections::HashMap, iter::FromIterator};

use super::*;
use crate::error::*;
use itertools::Itertools;
use std::collections::LinkedList;

type VarName = String;
type AttrValue = String;
type AttrName = String;

pub trait FromExpr: Sized {
    fn from_expr(e: Expr) -> AstResult<Self>;
}

impl FromExpr for Expr {
    fn from_expr(e: Expr) -> AstResult<Self> {
        Ok(e)
    }
}

#[derive(Debug, Eq, PartialEq)]
pub enum DefType {
    Widget,
}

impl FromExpr for DefType {
    fn from_expr(e: Expr) -> AstResult<Self> {
        if let Expr::Symbol(span, sym) = e {
            match sym.as_str() {
                "defwidget" => Ok(DefType::Widget),
                _ => Err(AstError::InvalidDefinition(Some(span))),
            }
        } else {
            Err(AstError::WrongExprType(Some(e.span()), ExprType::Symbol, e))
        }
    }
}

#[derive(Debug, Eq, PartialEq)]
pub struct Definitional<C, A> {
    def_type: DefType,
    name: String,
    attrs: HashMap<AttrName, A>,
    children: Vec<C>,
    span: Span,
}

impl<C: FromExpr, A: FromExpr> FromExpr for Definitional<C, A> {
    fn from_expr(e: Expr) -> AstResult<Self> {
        let span = e.span();
        spanned!(e.span(), {
            let list = e.as_list()?;
            let mut iter = itertools::put_back(list.into_iter());

            let def_type = DefType::from_expr(iter.next().or_missing(ExprType::Symbol)?)?;
            let name = iter.next().or_missing(ExprType::Symbol)?.as_symbol()?;
            let attrs = parse_key_values(&mut iter)?;
            let children = iter.map(|x| C::from_expr(x)).collect::<AstResult<Vec<_>>>()?;

            Definitional { span, def_type, name, attrs, children }
        })
    }
}

#[derive(Debug, Eq, PartialEq)]
pub struct Element<C, A> {
    name: String,
    attrs: HashMap<AttrName, A>,
    children: Vec<C>,
    span: Span,
}

impl<C: FromExpr, A: FromExpr> FromExpr for Element<C, A> {
    fn from_expr(e: Expr) -> AstResult<Self> {
        let span = e.span();
        spanned!(e.span(), {
            let list = e.as_list()?;
            let mut iter = itertools::put_back(list.into_iter());

            let name = iter.next().or_missing(ExprType::Str)?.as_symbol()?;
            let attrs = parse_key_values(&mut iter)?;
            let children = iter.map(C::from_expr).collect::<AstResult<Vec<_>>>()?;

            Element { span, name, attrs, children }
        })
    }
}

/// Parse consecutive `:keyword value` pairs from an expression iterator into a HashMap. Transforms the keys using the FromExpr trait.
fn parse_key_values<T: FromExpr, I: Iterator<Item = Expr>>(iter: &mut itertools::PutBack<I>) -> AstResult<HashMap<String, T>> {
    let mut data = HashMap::new();
    loop {
        match iter.next() {
            Some(Expr::Keyword(span, kw)) => match iter.next() {
                Some(value) => {
                    data.insert(kw, T::from_expr(value)?);
                }
                None => {
                    iter.put_back(Expr::Keyword(span, kw));
                    return Ok(data);
                }
            },
            Some(expr) => {
                iter.put_back(expr);
                return Ok(data);
            }
            None => return Ok(data),
        }
    }
}

#[cfg(test)]
mod test {

    use super::*;
    use insta;

    #[test]
    fn test() {
        let parser = parser::ExprParser::new();
        insta::with_settings!({sort_maps => true}, {
            insta::assert_debug_snapshot!(
                Element::<Expr, Expr>::from_expr(parser.parse("(box :bar 12 :baz \"hi\" foo (bar))").unwrap()).unwrap()
            );
            insta::assert_debug_snapshot!(
                Definitional::<Expr, Expr>::from_expr(parser.parse("(defwidget box (child) (child2))").unwrap()).unwrap()
            );
        });
    }
}
