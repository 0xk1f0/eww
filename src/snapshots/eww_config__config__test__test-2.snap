---
source: src/config.rs
expression: "Definitional::<Expr,\n               Expr>::from_expr(parser.parse(\"(defwidget box (child) (child2))\").unwrap()).unwrap()"

---
Definitional {
    def_type: Widget,
    name: "box",
    attrs: {},
    children: [
        List(
            15..22,
            [
                Symbol(
                    16..21,
                    "child",
                ),
            ],
        ),
        List(
            23..31,
            [
                Symbol(
                    24..30,
                    "child2",
                ),
            ],
        ),
    ],
    span: 1..10,
}
